File: 051/3.js
class Bookable {

    checkAvailability() {
      throw "checkAvailability method must be implemented";
    }
  }
  
  class Room extends Bookable {
    constructor(roomNumber, roomType) {
      super();
      this._roomNumber = roomNumber;
      this._roomType = roomType;
      this._isBooked = false;
    }
  

    getRoomNumber() {
      return this._roomNumber;
    }
  
    setRoomNumber(roomNumber) {
      this._roomNumber = roomNumber;
    }
  
  
    getRoomType() {
      return this._roomType;
    }
  
    setRoomType(roomType) {
      this._roomType = roomType;
    }
  

    getIsBooked() {
      return this._isBooked;
    }
  
    setIsBooked(isBooked) {
      this._isBooked = isBooked;
    }
  
    bookRoom() {
      if (!this._isBooked) {
        this._isBooked = true;
        console.log(`ห้อง ${this._roomNumber} ถูกจองเเล้ว`);
      } else {
        console.log(`ห้อง ${this._roomNumber} มีการจองไว้แล้ว.`);
      }
    }
  
    cancelBooking() {
      if (this._isBooked) {
        this._isBooked = false;
        console.log(`ห้อง ${this._roomNumber} การจองถูกยกเลิก`);
      } else {
        console.log(`ห้อง ${this._roomNumber} ไม่ได้จองเเล้ว`);
      }
    }
  
    displayDetails() {
      let status = this._isBooked ? "ห้อง" : "มีอยู่";
      console.log(`ห้องหมายเลข: ${this._roomNumber}, ประเภทห้อง: ${this._roomType}, สถานะ: ${status}`);
    }
  
    checkAvailability() {
      return !this._isBooked;
    }
  }
  
  class DeluxeRoom extends Room {
    constructor(roomNumber) {
      super(roomNumber, "Deluxe");
    }
  
    displayDetails() {
      super.displayDetails();
      console.log("ห้องนี้เป็นห้องดีลักซ์พร้อมสิ่งอำนวยความสะดวกระดับพรีเมียม");
    }
  }
  
  class StandardRoom extends Room {
    constructor(roomNumber) {
      super(roomNumber, "Standard");
    }
  
    displayDetails() {
      super.displayDetails();
      console.log("ห้องนี้เป็นห้องมาตรฐานพร้อมสิ่งอำนวยความสะดวกขั้นพื้นฐาน");
    }
  }
  
  // Demonstration of Hotel Booking System
  const room1 = new DeluxeRoom(101);
  const room2 = new StandardRoom(102);
  const room3 = new DeluxeRoom(103);
  
  room1.displayDetails();
  room1.bookRoom();
  room1.displayDetails();
  room1.cancelBooking();
  room1.displayDetails();
  
  console.log("\n");
  room2.displayDetails();
  room2.bookRoom();
  room2.displayDetails();
  
  console.log("\n");
  room3.displayDetails();
  room3.bookRoom();
  room3.cancelBooking();
  room3.displayDetails();
  

File: 051/3.py
from abc import ABC, abstractmethod

class Bookable(ABC):
    @abstractmethod
    def checkAvailability(self):
        pass

class Room(Bookable):
    def __init__(self, roomNumber, roomType):
        self.__roomNumber = roomNumber
        self.__roomType = roomType
        self.__isBooked = False

    
    def getRoomNumber(self):
        return self.__roomNumber

    def setRoomNumber(self, roomNumber):
        self.__roomNumber = roomNumber

    
    def getRoomType(self):
        return self.__roomType

    def setRoomType(self, roomType):
        self.__roomType = roomType

    
    def getIsBooked(self):
        return self.__isBooked

    def setIsBooked(self, isBooked):
        self.__isBooked = isBooked

    def bookRoom(self):
        if not self.__isBooked:
            self.__isBooked = True
            print(f"ห้อง {self.__roomNumber} ถูกจองเเล้ว")
        else:
            print(f"ห้อง {self.__roomNumber} มีการจองไว้แล้ว.")

    def cancelBooking(self):
        if self.__isBooked:
            self.__isBooked = False
            print(f"ห้อง {self.__roomNumber} การจองถูกยกเลิก")
        else:
            print(f"ห้อง {self.__roomNumber} ไม่ได้จองเเล้ว")

    def displayDetails(self):
        status = "ห้อง" if self.__isBooked else "มีอยู่"
        print(f"ห้องหมายเลข: {self.__roomNumber}, ประเภทห้อง: {self.__roomType}, สถานะ: {status}")

    def checkAvailability(self):
        return not self.__isBooked

class DeluxeRoom(Room):
    def __init__(self, roomNumber):
        super().__init__(roomNumber, "Deluxe")

    def displayDetails(self):
        super().displayDetails()
        print("ห้องนี้เป็นห้องดีลักซ์พร้อมสิ่งอำนวยความสะดวกระดับพรีเมียม")

class StandardRoom(Room):
    def __init__(self, roomNumber):
        super().__init__(roomNumber, "Standard")

    def displayDetails(self):
        super().displayDetails()
        print("ห้องนี้เป็นห้องมาตรฐานพร้อมสิ่งอำนวยความสะดวกขั้นพื้นฐาน")


room1 = DeluxeRoom(101)
room2 = StandardRoom(102)
room3 = DeluxeRoom(103)

room1.displayDetails()
room1.bookRoom()
room1.displayDetails()
room1.cancelBooking()
room1.displayDetails()

print("\n")
room2.displayDetails()
room2.bookRoom()
room2.displayDetails()

print("\n")
room3.displayDetails()
room3.bookRoom()
room3.cancelBooking()
room3.displayDetails()
